_source/_posts/2021-11-08-k8s-api-server-oidc.md

```diff
@@ -0,0 +1,562 @@
+  resources: ["pods", "services"]
+  verbs: ["get", "watch", "list"]
+EOF
+```
+
+Now we can bind the role to the second group we created in Okta.
+
+```bash
+kubectl apply -f - <<EOF
+kind: RoleBinding
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: oidc-cluster-user
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: restricted-user
+subjects:
+- kind: Group
+  name: k8s-restricted-users
+EOF
+```
+
+That's it! The cluster is now ready for some OIDC action.
+
+## Connecting to the cluster using kubectl
+
+Before we can go ahead and test this out, we need to do some setup for kubectl so that it knows how to do OIDC authentication. We need to install [kubelogin](https://github.com/int128/kubelogin) plugin for this. Go ahead and install it using any of the following commands.
+
+```bash
+# Homebrew (macOS and Linux)
+brew install int128/kubelogin/kubelogin
+
+# Krew (macOS, Linux, Windows and ARM)
+kubectl krew install oidc-login
+```
+
+The plugin enables OIDC login capability for kubectl. Let's test it out first. Run the following command, make sure to replace `k8s_oidc_issuer_url` and `k8s_oidc_client_id` with what you saved earlier during Okta set up.
+
+```bash
+kubectl oidc-login setup --oidc-issuer-url=<k8s_oidc_issuer_url> --oidc-client-id=<k8s_oidc_client_id>
+```
+
+This should produce an output that looks like this:
+
+```
+authentication in progress...
+Opening in existing browser session.
+
+## 2. Verify authentication
+
+You got a token with the following claims:
+
+{
+  "sub": "00u28lpcoheT48W5A5d7",
+  "ver": 1,
+  "iss": "<k8s_oidc_issuer_url>",
+  "aud": "<k8s_oidc_client_id>",
+  "iat": 1634657188,
+  "exp": 1634660788,
+  "jti": "ID.AmO7M7Z4xh_rSN4IHVuQEfJ4QuoADOWHSrVHuNyn9so",
+  "amr": [
+    "pwd"
+  ],
+  "idp": "0oa28lrldhsBglcC35d7",
+  "nonce": "2Rsv7E84vo_K14J6qm9IgKjNsCUxb4Hc16qgzx9CKlk",
+  "auth_time": 1634649927,
+  "at_hash": "veXQkvqTjCl70OkZnKZVyA",
+  "groups": [
+    "k8s-admins"
+  ]
+}
+
+```
+
+Now let us update kubectl configuration to add an OIDC user:
+
+```bash
+kubectl config set-credentials oidc-user \
+--exec-api-version=client.authentication.k8s.io/v1beta1 \
+--exec-command=kubectl \
+--exec-arg=oidc-login \
+--exec-arg=get-token \
+--exec-arg=--oidc-issuer-url=<k8s_oidc_issuer_url> \
+--exec-arg=--oidc-client-id=<k8s_oidc_client_id> \
+--exec-arg=--oidc-extra-scope="email offline_access profile openid"
+```
+
+Now we can use this user to authenticate to the cluster via kubectl.
+
+```bash
+kubectl get pods --user=oidc-user -n default
+```
+
+This should open a browser window to authenticate you. You can use your Okta account credentials to log in.
+
+You can play around with different users in different groups to test out the RBAC configuration we did. Users who are not in the `k8s-admins` group in Okta should not be able to access any resource. Users who are only in the `restricted-users` group should only be able to see pods and services.
```

Comment:
```suggestion
Test out this RBAC configuration by playing around with different users in different groups. Users who are not in the `k8s-admins` group in Okta should not be able to access any resource. Users who are only in the `restricted-users` group should only be able to see pods and services.
```
sentence variation to avoid repetitive sentence structure...